### PR TITLE
Add tuple, arrays of tuples, and dynamic tuple support

### PR DESCRIPTION
**Pull request by eugenetian**
This Pull Request introduces enhanced functionality to our existing Ethereum Decoder, allowing it to correctly interpret and handle tuple types and arrays of tuple types. This addition greatly improves the accuracy and richness of the data we can extract and use from Ethereum transactions.

**Key Changes:**

**Tuple Decoding:** The decoder can now accurately parse tuple types in Ethereum transactions. This is a critical enhancement as tuples allow for more complex data structures and can contain multiple different types of data within one parameter.

**Array of Tuples Decoding:** The extension of the decoder to handle arrays of tuples marks a significant improvement in our decoding capabilities. As Ethereum smart contracts and transactions become more complex, arrays of tuples become more commonly used, especially in DeFi protocols. This feature ensures our ability to stay compatible with the latest Ethereum transaction standards.

**Dynamic Type Handling:** The update also includes improvements in how we handle dynamic types. This includes strings, bytes, and arrays, ensuring that we can accurately decode the complete parameter list of transactions that utilize these types.

**Regex Library Utilization:** We have introduced the regex library to ensure efficient and accurate extraction of necessary information from the types and overall decoding process.

**Code Refactoring:** The code has undergone a major refactoring to increase readability and maintainability. This will also make future enhancements easier to implement.

Tests have been conducted to verify the correct operation of these changes. With this update, we anticipate being able to handle more sophisticated Ethereum transaction data structures, thereby delivering more accurate and detailed data to our users.

### CODE REVIEW

File: macros/decode_logs.sql [hunk 1]
```diff
@@ -19,18 +21,41 @@ def decode_logs(abi, data, topics):
         decoded_result = {}
         data_offset = 0
         decode_success = True
-        for input_obj in json_object['inputs']:
-            if input_obj['indexed'] == True:
-                decode_single_res, decoded_successfully = decode_single(topics_array[topics_index][2:], input_obj['type'], 0)
+        if data == "":
+            for input_obj in json_object['inputs']:
+                print(topics_array[topics_index][2:])
```
Comment (eunseochoi) [inline macros/decode_logs.sql:26]: 🔪

**Review by eunseochoi — APPROVED** [review]

**Review by curlymuse — COMMENTED** [review]
General comments:

- Don't use `else` or `elif` unless it's absolutely necessary. 9 times out of 10, you can avoid `else` with a `return`, a `continue` or a `break` and the code will be way cleaner and more readable.
- I'd suggest using `match`/`case` (afaik this is the Python equivalent of `switch`) in the main function, then breaking the logic for each type into its own smaller function. A 150+ line function is really difficult to read and maintain and is rarely necessary
- Since this logic is very granular and close to the metal, leave some comments summarizing what is being done in each block of steps. Remember: this is open source code. It needs to be in a condition for others to collaborate on it, which means it should be well documented and ready for collaborators.

File: macros/decode_logs.sql (hunk 1 above)
Comment (eugenetian) [reply macros/decode_logs.sql]: oops LOL

**Comment by eugenetian** [comment]
> General comments:
> 
> * Don't use `else` or `elif` unless it's absolutely necessary. 9 times out of 10, you can avoid `else` with a `return`, a `continue` or a `break` and the code will be way cleaner and more readable.
> * I'd suggest using `match`/`case` (afaik this is the Python equivalent of `switch`) in the main function, then breaking the logic for each type into its own smaller function. A 150+ line function is really difficult to read and maintain and is rarely necessary
> * Since this logic is very granular and close to the metal, leave some comments summarizing what is being done in each block of steps. Remember: this is open source code. It needs to be in a condition for others to collaborate on it, which means it should be well documented and ready for collaborators.

Makes sense, I'll create a follow up PR to make the code more readable

**Pull request merged**